### PR TITLE
Add proposal about supporting git repo in Helm Broker

### PR DESCRIPTION
**Pull request by mszostok**
**Description**

Changes proposed in this pull request:

- Add proposal about supporting git repo in Helm Broker

**Related issue(s)**

- https://github.com/kyma-project/kyma/issues/4985

### CODE REVIEW

File: components/helm-broker/docs/proposals/addon-git-support.md [hunk 1]
```diff
@@ -0,0 +1,163 @@
+# Support for loading Addons from Git repository
+
+Created on 2019-07-24 by Mateusz Szostok ([@mszostok](https://github.com/mszostok)).
+
+This document describes options for extending the ClusterAddonsConfiguration/AddonsConfiguration for supporting loading Addons from the Git repository.
+
+## Motivation
+
+Helm Broker fetches addons listed in the `index.yaml` file from remote HTTPS servers. Only zipped bundles can be fetched. This solution generates problems for end-user as it's not easy to always have dedicated HTTPS server which serves zipped bundles. A most common scenario is to have a git repository with unzipped bundles and index YAMLs. This document describes possible and accepted solutions for achieving that goal.
```
Comment (PK85) [inline components/helm-broker/docs/proposals/addon-git-support.md]: Please mentioned at the end, that this syntax in addition opens a doors to other types of storage, like s3, mercurial etc.

File: components/helm-broker/docs/proposals/addon-git-support.md [hunk 2]
```diff
@@ -0,0 +1,163 @@
...
+
+## Accepted solution
```
Comment (PK85) [inline components/helm-broker/docs/proposals/addon-git-support.md]: Please mentioned at the end, that this syntax in addition opens a doors to other types of storage, like s3, mercurial etc.

File: components/helm-broker/docs/proposals/addon-git-support.md [hunk 3]
```diff
@@ -0,0 +1,163 @@
...
+
+```yaml
+apiVersion: addons.kyma-project.io/v1alpha1
+kind: ClusterAddonsConfiguration
+metadata:
+  name: addons-cfg-sample
+spec:
+  repositories:
+    # Use SSH protocol with path to index YAML
+    - url: "git::git@github.com:kyma-project/bundles.git//index.yaml"
```
Comment (PK85) [inline components/helm-broker/docs/proposals/addon-git-support.md]: I understand that https repositories, will have URLs as before? E.g. https://github.com/kyma-project/bundles/releases/download/0.6.0/index-testing.yaml

File: components/helm-broker/docs/proposals/addon-git-support.md [hunk 4]
```diff
@@ -0,0 +1,163 @@
...
+</details>
+
+
+## Support for authorization
+
+Support for authorization is considered to be added in future. 
+
+### Proposal
+
+User can specify the reference to secret where credentials are stored. In URL, user needs to specify query parameters with key-value pairs. As value set the name of the key from Secret which constrains token data. Helm Broker is reading ClusterAddonsConfiguration and render the url by replacing the **{KEY_NAME_FROM_SECRET_PLACEHOLDER}** with proper values extracted from secret referenced by the `authSecretRef` field.
```
Comment (klaudiagrz) [inline components/helm-broker/docs/proposals/addon-git-support.md]:
```suggestion
In the future, we want to allow you to specify the reference to a Secret in which you store credentials. In URL, you must specify query parameters with key-value pairs. As a value, set the name of the key from a Secret which constrains token data. Helm Broker is reading ClusterAddonsConfiguration and renders the URL by replacing the **{KEY_NAME_FROM_SECRET_PLACEHOLDER}** with proper values extracted from Secret referenced by the **authSecretRef` field.
```

Comment (klaudiagrz) [reply components/helm-broker/docs/proposals/addon-git-support.md]: Let's discuss.

File: components/helm-broker/docs/proposals/addon-git-support.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (mszostok) [reply components/helm-broker/docs/proposals/addon-git-support.md]: yes

Comment (mszostok) [reply components/helm-broker/docs/proposals/addon-git-support.md]: exactly the https/http repos should be specified in the same way as previously. go-getter also supports that

proper example added

File: components/helm-broker/docs/proposals/addon-git-support.md [hunk 5]
```diff
@@ -0,0 +1,166 @@
+# Support fetching addons from Git repository
+
+Created on 2019-07-24 by Mateusz Szostok ([@mszostok](https://github.com/mszostok)).
+
+This document describes options to extend the ClusterAddonsConfiguration and AddonsConfiguration custom resources (CRs) so that they support fetching addons from any Git repository.
+
+## Motivation
+
+Helm Broker fetches addons listed in the `index.yaml` file and exposed as remote HTTPS servers, which means that only zipped addons can be fetched. This solution generates problems for end-users, as it's not always easy to have a dedicated HTTPS server that serves zipped addons. The most common scenario is to have a Git repository with unzipped addons and index yaml files. This document describes all possible solutions for achieving that goal.
```
Comment (klaudiagrz) [inline components/helm-broker/docs/proposals/addon-git-support.md]:
```suggestion
Helm Broker fetches addons listed in the `index.yaml` file and exposed as remote HTTPS servers, which means that only zipped addons can be fetched. This solution generates problems for end-users, as it's not always easy to have a dedicated HTTPS server that serves zipped addons. The most common scenario is to have a Git repository with unzipped addons and index.yaml files. This document describes all possible solutions for achieving that goal.
```

File: components/helm-broker/docs/proposals/addon-git-support.md [hunk 6]
```diff
@@ -0,0 +1,166 @@
...
+
+## Accepted solution
+
+You can specify Git repositories by adding the special `git::` prefix to the URL address. After this prefix, you can provide any valid Git URL with one of the protocols supported by Git. This solution is inspired by the [Terraform implementation](https://www.terraform.io/docs/modules/sources.html#github). Using the [hashicorp go-getter](https://github.com/hashicorp/go-getter) allows us to easily add new supported protocols such as Mercurial, S3, GCS etc.
```
Comment (klaudiagrz) [inline components/helm-broker/docs/proposals/addon-git-support.md]:
```suggestion
You can specify Git repositories by adding the special `git::` prefix to the URL address. After this prefix, you can provide any valid Git URL with one of the protocols supported by Git. This solution is inspired by the [Terraform implementation](https://www.terraform.io/docs/modules/sources.html#github). Using the [hashicorp go-getter](https://github.com/hashicorp/go-getter) allows us to easily add new supported protocols, such as Mercurial, S3, GCS, etc.
```

File: components/helm-broker/docs/proposals/addon-git-support.md [hunk 7]
```diff
@@ -0,0 +1,166 @@
...
+
+>**NOTE:** This implementation uses the Git binary. It check-outs the full repository but depth can be specified.
+
+</p>
+</details>
+
+<details><summary>Cons</summary>
+<p>
+
+- User must know the new pattern, which is not obvious
```
Comment (klaudiagrz) [inline components/helm-broker/docs/proposals/addon-git-support.md]:
```suggestion
- User must know the new pattern, which is not obvious.
```

File: components/helm-broker/docs/proposals/addon-git-support.md [hunk 8]
```diff
@@ -0,0 +1,166 @@
...
+    - type: git
+      ref: master # branch/tag/commit
+      url: git@github.com:kyma-project/bundles.git # all git protocol supported
+      indexPath: bundles/bundles/index.yaml
+``` 
+
+<details><summary>Pros</summary>
+<p>
+
+- Support all git repositories
```
Comment (klaudiagrz) [inline components/helm-broker/docs/proposals/addon-git-support.md]:
```suggestion
- Support all Git repositories
```

File: components/helm-broker/docs/proposals/addon-git-support.md [hunk 9]
```diff
@@ -0,0 +1,166 @@
...
+- API is self-describing
+
+</p>
+</details>
+
+<details><summary>Cons</summary>
+<p>
+
+- Migration job needed
+- Changes in UI and console-backend-service needed
```
Comment (klaudiagrz) [inline components/helm-broker/docs/proposals/addon-git-support.md]:
```suggestion
- Changes in UI and Console Backend Service needed
```

File: components/helm-broker/docs/proposals/addon-git-support.md [hunk 10]
```diff
@@ -0,0 +1,166 @@
...
+- Link is not "clickable". In UI you are not able to click directly the link and go to a given repository.
```
Comment (klaudiagrz) [inline components/helm-broker/docs/proposals/addon-git-support.md]:
```suggestion
- Link is not clickable. You cannot click the link directly in UI and go to a given repository.
```

File: components/helm-broker/docs/proposals/addon-git-support.md [hunk 11]
```diff
@@ -0,0 +1,166 @@
...
+</details>
+
+
+## Future implementations
+
+We consider adding support for authorization in the future. 
+
+### Support for authorization
+
+In the future, we want to allow you to specify the reference to a Secret in which you store credentials. In URL, you must specify query parameters with key-value pairs. As a value, set the name of the key from a Secret which constrains token data. Helm Broker is reading ClusterAddonsConfiguration and renders the URL by replacing the **{KEY_NAME_FROM_SECRET_PLACEHOLDER}** with proper values extracted from Secret referenced by the **authSecretRef` field.
```
Comment (klaudiagrz) [inline components/helm-broker/docs/proposals/addon-git-support.md]:
```suggestion
In the future, we want to allow you to specify the reference to a Secret in which you store credentials. In URL, you must specify query parameters with key-value pairs. As a value, set the name of the key from a Secret which contains token data. Helm Broker fetches ClusterAddonsConfigurations and renders URLs by replacing the **{KEY_NAME_FROM_SECRET_PLACEHOLDER}** with proper values extracted from Secrets referenced by the **authSecretRef** field.
```

File: components/helm-broker/docs/proposals/addon-git-support.md [hunk 12]
```diff
@@ -0,0 +1,166 @@
...
+  repositories:
+    # Use Git SSH protocol with a path to index.yaml
+    - url: "git::git@github.com:kyma-project/addons.git//index.yaml"
+    # Use Git HTTPS protocol with a path to index.yaml 
+    - url: "git::https://github.com/kyma-project/addons.git//addons/index.yaml"
+    # Use Git HTTPS protocol with a path to index.yaml and branch/tag version 
+    - url: "git::https://github.com/kyma-project/addons.git//addons/index.yaml?ref=1.2.0"
+    # Use unprefixed github.com URLs. They are automatically interpreted as Git repository sources. 
+    - url: "github.com/kyma-project/addons//addons/index.yaml?ref=1.2.0"
+    # Use HTTPS protocol (server which serves static content), should be defined as previously. Nothing changed.
```
Comment (klaudiagrz) [inline components/helm-broker/docs/proposals/addon-git-support.md]:
```suggestion
    # Use HTTPS protocol (server which serves static content) defined the way it was implemented so far.
```

**Pull request merged**